executor/src/witgen/jit/function_cache.rs

```diff
@@ -0,0 +1,133 @@
+use std::collections::BTreeMap;
+
+use bit_vec::BitVec;
+use powdr_number::FieldElement;
+
+use crate::witgen::{
+    data_structures::finalizable_data::{ColumnLayout, CompactDataRef},
+    jit::affine_symbolic_expression::Effect,
+    machines::{LookupCell, MachineParts},
+    EvalError, FixedData, MutableState, QueryCallback,
+};
+
+use super::{
+    block_machine_processor::BlockMachineProcessor,
+    compiler::{compile_effects, WitgenFunction},
+    variable::Variable,
+};
+
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+struct CacheKey {
+    identity_id: u64,
+    known_args: BitVec,
+}
+
+pub struct FunctionCache<'a, T: FieldElement> {
+    /// The processor that generates the JIT code
+    processor: BlockMachineProcessor<'a, T>,
+    /// The cache of JIT functions. If the entry is None, we attempted to generate the function
+    /// but failed.
+    witgen_functions: BTreeMap<CacheKey, Option<WitgenFunction<T>>>,
```

Comment:
HashMap?

Reply:
When would you use which? I tend to always use `BTreeMap`. I *think* it's faster for a small number of elements? Also, performance is deterministic. But I guess in practice there is not a big difference either way?

Reply:
I assume hashing such a key is faster than doing 3-4 comparisons, but I might be wrong.